### PR TITLE
add `let_underscore_lock` lint

### LINKED ISSUE #1574: Warn on `let _ = x.lock();`
See #40096.

### PR DESCRIPTION
**Pull request by Areredify**
closes #1574
changelog: add `let_underscore_lock` lint

I am not entirely sure about my docs/messages wording here, improvements are welcome

### CODE REVIEW

**Comment by JohnTitor** [comment]
I think it'd be nice to separate test files.

**Comment by Areredify** [comment]
Oh, and I forgot to ask your input about this: is there a good reason to not lint `let _ = <expr with MutexGuard/RWGuard ty>`?

**Comment by flip1995** [comment]
> is there a good reason to not lint let _ = <expr with MutexGuard/RWGuard ty>?

I cannot come up with one. When a guard is produced in any way, dropping it immediately is probably a mistake.

This can be nicely checked with [`TyS::walk`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.walk). 

```rust
expr_ty(init_expr).walk().any(|ty| GUARD_TYPES.any(|gty| match_ty(gty, ty)));
```

File: clippy_lints/src/let_underscore.rs [hunk 1]
```diff
@@ -43,22 +71,37 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for LetUnderscore {
...
+                if_chain! {
+                    if let ExprKind::MethodCall(_, _, _) = init.kind;
+                    let method_did = cx.tables.type_dependent_def_id(init.hir_id).unwrap();
+                    if LOCK_METHODS_PATHS.iter().any(|path| match_def_path(cx, method_did, path));
+                    then {
+                        span_lint_and_help(
+                            cx,
+                            LET_UNDERSCORE_LOCK,
+                            stmt.span,
+                            "non-binding let on an a synchronization lock",
```
Comment (flip1995) [inline clippy_lints/src/let_underscore.rs]:
```suggestion
                            "non-binding let on a synchronization lock",
```

**Comment by Areredify** [comment]
I just thought, someone might use `let _ = mutex_lock` to drop the lock prematurely. Maybe we should lint all the functions that return a guard.

**Comment by flip1995** [comment]
> I just thought, someone might use let _ = mutex_lock to drop the lock prematurely.

In that case, I think `drop(mutex_lock)` would be way better.

**Comment by Areredify** [comment]
@flip1995 from the documentation it seems like `TyS::walk` walks generic parameters on a type recursively, why is it relevant here?

**Comment by flip1995** [comment]
`Result<_Guard, _>` is pretty usual, I think. And we also want to detect this `Guard` types.

**Comment by Areredify** [comment]
Oh, that's what you mean, gotcha

File: clippy_lints/src/let_underscore.rs [hunk 2]
```diff
@@ -30,7 +30,35 @@ declare_clippy_lint! {
...
+    /// let _lock = mutex.lock();
+    /// ```
+    pub LET_UNDERSCORE_LOCK,
+    correctness,
+    "non-binding let on a synchronization lock"
+}
+
+declare_lint_pass!(LetUnderscore => [LET_UNDERSCORE_MUST_USE, LET_UNDERSCORE_LOCK]);
+
+const LOCK_METHODS_PATHS: [&[&str]; 3] = [&paths::MUTEX_LOCK, &paths::RWLOCK_READ, &paths::RWLOCK_WRITE];
```
[the remark this answers is not in the document]
Comment (Areredify) [reply clippy_lints/src/let_underscore.rs]: We added it :+1:

File: clippy_lints/src/let_underscore.rs [hunk 3]
```diff
@@ -43,8 +76,17 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for LetUnderscore {
...
-                   span_lint_and_help(
+                let check_ty = |ty| SYNC_GUARD_PATHS.iter().any(|path| match_type(cx, ty, path));
+                if cx.tables.expr_ty(init).walk().any(check_ty) {
+                        span_lint_and_help(
+                            cx,
+                            LET_UNDERSCORE_LOCK,
+                            stmt.span,
+                            "non-binding let on a synchronization lock",
+                            "consider using an underscore-prefixed named binding or dropping explicitly with `std::mem::drop`"
+                        )
```
Comment (flip1995) [inline clippy_lints/src/let_underscore.rs]: double indentation

Comment (Areredify) [reply clippy_lints/src/let_underscore.rs]: Why doesn't fmt do that?

Comment (flip1995) [reply clippy_lints/src/let_underscore.rs]: rustfmt currently only formats in some macros. The `if_chain` macro is too complex, for rustfmt to format the code in it reliably.

**Comment by flip1995** [comment]
@bors r+ 

Travis has currently problems, so I marked it as `S-waiting-on-rustup` to merge this as soon as travis works again.

**Comment by mleduque** [comment]
I see this in my code now.
There are
`let _ = SHARED_OBJECT.lock().map(|content| { ... });`

Is that wrong ?

**Comment by flip1995** [comment]
Uh, no that is ok. @Areredify I think we found the reason, why we should have only linted on `lock` functions, instead of the types.

**Comment by flip1995** [comment]
Hm thinking more about this: With your `map(..)` call, `content` is not the thing inside the `Mutex`, but the `MutexGuard` itself. So your type is probably (?):

```rust
let _ = SHARED_OBJECT.lock().map(|content: MutexGuard<impl Iterator<Item = _>>| { ... });
```

From the `lock` documentation:

> If another user of this mutex panicked while holding the mutex, then this call will return an error once the mutex is acquired.

So `map`ping the result of the `lock()` call only really makes sense, if you do some error handling (other thread `panic`ed) for the above case in there. But I wouldn't do error handling in a `map` call, but rather in a `match _.lock() { Ok(_) => { ... }, Err(_) => panic!("..."),`. Or you do 

```rust
SHARED_OBJECT.lock().expect("..").map(|content: <Iterator>::Item>| { ... })
```

**Comment by mleduque** [comment]
I had to check, I didn't write the code, it was just pointed to me by clippy (that's totally the point of clippy, eh! Thanks by the way).

It seems the error case is actually ignored, like I wrote above. This is probably a bug in itself (or maybe not, I'll have to check the _intent_ with the original authors, I guess).

**Comment by flip1995** [comment]
> I'll have to check the intent with the original authors

If it should be a bug of this lint, feel free to open an issue about it :+1:

**Comment by mleduque** [comment]
I don't think I'm qualified enough to decide if there is a bug or not :D

I also found some
`let _ = SHARED_OBJECT.lock().map_or_else(|error| {...}, |content| { ... })`

So I think some of the `map` ones may actually be as intended (but possibly not all of them, but I haven't yet had the authors input on the subject).

**Pull request merged**